### ISSUE TITLE
net/haproxy: add support for TLS 1.3

### ISSUE THREAD

**Issue by fraenki** [opnsense/plugins#790]
- [x] requires HAProxy 1.9+ (it [added support](https://www.mail-archive.com/haproxy@formilux.org/msg32143.html) for TLS 1.3 ciphers, #1089)
- [x] requires OpenSSL 1.1.1+ or LibreSSL with TLS 1.3 support (ETA OPNsense 20.7)
- [x] add TLS 1.3 GUI options for HAProxy plugin

**Comment by fabianfrz** [comment]
@fraenki In nginx I've enabled TLS 1.3 already but since the crypto library does not support it, it won't be usable. See libressl-portable/portable#228

**Comment by fabianfrz** [comment]
@fichtner My VPS is now running TLS 1.3 so it is probably not a big step to integrate it now.
![unbenannt](https://user-images.githubusercontent.com/3713914/46913664-fa384780-cf91-11e8-8674-a2c559c79717.png)

This is how it currently looks on my Arch machine:

* extra/nginx 1.14.0-1
* core/openssl 1.1.1-1
* core/openssl-1.0 1.0.2.p-1 (was kept for legacy apps but was safe to remove from the server, on my host machine it is still required like you can see below in the packman output)

```
Benötigt von             : john  wpa_supplicant
```

My host has squid 4, nginx and lots of other stuff, the server has postfix, nginx etc.
So all major services used by OPNsense are also working with the up to date openssl version but I don't know if wpa_supplicant is also in OPNsense.

**Comment by fichtner** [comment]
we won't have this until we have 1.1.1 in openssl ports as the *default*. FreeBSD 12 is experimenting with last minute inclusion of 1.1.1 which doesn't go as well as it should. We'll just have to wait...

**Comment by fichtner** [comment]
PS: LibreSSL and TLS 1.3 is currently unlikely to land in the new 2.8 so that would mean 2.9 which is easily 6 months away.

**Comment by fabianfrz** [comment]
@fichtner shouldn't that be no issue because FreeBSD has its own openssl binary?

**Comment by fabianfrz** [comment]
btw. 1.1.1 is available as openssl-111

**Comment by fichtner** [comment]
True, but I have no intention of potentially destabilising an area that currently works well. Build breaks often enough with supposedly working release software and there's 11.2 to tackle, too.

**Comment by fabianfrz** [comment]
@fichtner  19.7 roadmap then?

**Comment by irgendwr** [comment]
Any update on this? Support for TLS 1.3 would be really nice.

**Comment by mimugmail** [comment]
This needs Openssl 1.1.1 which requires FreeBSD 12.0 .. wont come before 20.1

**Comment by fichtner** [comment]
And it won‘t work with LibreSSL so far.

**Comment by Taomyn** [comment]
Are we any closer to getting TLS 1.3 now? I believe it's possible for the admin GUI according to this thread on the forums: https://forum.opnsense.org/index.php?topic=15701.0

**Comment by fraenki** [comment]
On OPNsense 20.7 support for TLS 1.3 is available:

```
root@opnsense:~ # haproxy -vv|grep -i tls
OpenSSL library supports TLS extensions : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
```

I'll add the required GUI options.

**Comment by fraenki** [comment]
This will be available in the upcoming os-haproxy 2.25. It seems to work as expected:

```
$ openssl s_client -connect opnsense.example.com:8443 -tls1_3
...
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
```

**Comment by fichtner** [comment]
Note this is Client-only for LibreSSL...
